x-pack/legacy/plugins/apm/public/components/app/Settings/CustomizeUI/CustomLink/__test__/helper.test.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import {
+  convertFiltersToArray,
+  convertFiltersToObject,
+  getSelectOptions,
+  filterSelectOptions
+} from '../CustomLinkFlyout/helper';
+import { CustomLink } from '../../../../../../../../../../plugins/apm/server/lib/settings/custom_link/custom_link_types';
+
+describe('Custom link helper', () => {
+  describe('convertFiltersToArray', () => {
+    it('returns array of tuple when custom link not defined', () => {
+      expect(convertFiltersToArray()).toEqual([['', '']]);
+    });
+    it('returns filters as array', () => {
+      expect(
+        convertFiltersToArray({
+          'service.name': 'foo',
+          'transaction.type': 'bar'
+        } as CustomLink)
+      ).toEqual([
+        ['service.name', 'foo'],
+        ['transaction.type', 'bar']
+      ]);
+    });
+    it('returns empty when no filter is added', () => {
+      expect(
+        convertFiltersToArray({
+          label: 'foo',
+          url: 'bar'
+        } as CustomLink)
+      ).toEqual([['', '']]);
+    });
+  });
+
+  describe('convertFiltersToObject', () => {
+    it('returns undefined when any filter is added', () => {
+      expect(convertFiltersToObject([['', '']])).toBeUndefined();
+    });
+    it('removes uncompleted filters', () => {
+      expect(
+        convertFiltersToObject([
+          ['service.name', ''],
+          ['', 'foo'],
+          ['transaction.type', 'bar']
+        ])
+      ).toEqual({ 'transaction.type': ['bar'] });
+    });
+    it('splits the value by comma', () => {
+      expect(
+        convertFiltersToObject([
+          ['service.name', 'foo'],
+          ['service.environment', 'foo, bar'],
+          ['transaction.type', 'foo, '],
+          ['transaction.name', 'foo,']
+        ])
+      ).toEqual({
+        'service.name': ['foo'],
+        'service.environment': ['foo', 'bar'],
+        'transaction.type': ['foo'],
+        'transaction.name': ['foo']
+      });
+    });
+  });
+
+  describe('getSelectOptions', () => {
+    it('returns all option', () => {
+      expect(getSelectOptions([['', '']], 0)).toEqual(filterSelectOptions);
```

Comment:
What about changing `getSelectOptions` so its signature is `(filter, idx) => boolean` (like Array.prototype.filter)

So in [here](https://github.com/elastic/kibana/blob/156066dc6fa512169f1af4244f5c4cdefb61be14/x-pack/legacy/plugins/apm/public/components/app/Settings/CustomizeUI/CustomLink/CustomLinkFlyout/FiltersSection.tsx#L83)  you can do:
```js
filters.filter(getSelectOptions).map(filter, idx) =>
```

Reply:
.. looking at this again, I'm not sure my comment above makes sense. But it seems like `getSelectOptions` could be improved. It's invoked inside a map iteration and does itself two nested iterations. Seems a little wasteful. It's not so much the perf I'm worried about but I think it could be simplified.

Reply:
@sqren I will take a look

Reply:
@sqren WDYT https://github.com/elastic/kibana/pull/59648/commits/23402d7ce2767338da8561dacb6a875e8380128a#diff-87ac5566f7fceb0cd9e819016d4c7c8bR102-R111?